packages/components/src/components/CodeEditor.tsx

```diff
@@ -1,76 +1,201 @@
-
-  // this is necessary because AceEditor binds commands on mount, see https://github.com/securingsincity/react-ace/issues/684
-  const onSubmitRef = useRef<typeof onSubmit | null>(null);
-  onSubmitRef.current = onSubmit;
-
-  const editorEl = useRef<AceEditor | null>(null);
-
-  return (
-    <AceEditor
-      ref={editorEl}
-      value={value}
-      mode="golang"
-      theme="github"
-      width="100%"
-      fontSize={14}
-      height={String(height) + "px"}
-      minLines={oneLine ? lineCount : undefined}
-      maxLines={oneLine ? lineCount : undefined}
-      showGutter={showGutter}
-      highlightActiveLine={false}
-      showPrintMargin={false}
-      onChange={onChange}
-      name={id}
-      editorProps={{
-        $blockScrolling: true,
-      }}
-      setOptions={{}}
-      commands={[
-        {
-          name: "submit",
-          bindKey: { mac: "Cmd-Enter", win: "Ctrl-Enter" },
-          exec: () => onSubmitRef.current?.(),
-        },
-      ]}
-      markers={errorLocations?.map((location) => ({
-        startRow: location.start.line - 1,
-        startCol: location.start.column - 1,
-        endRow: location.end.line - 1,
-        endCol: location.end.column - 1,
-        className: "ace-error-marker",
-        type: "text",
-      }))}
-    />
+  const editor = useRef<HTMLDivElement>(null);
+  const editorView = useRef<EditorView | null>(null);
+  const languageSupport = useMemo(squiggle, []);
+  const state = useMemo(
+    () =>
+      EditorState.create({
+        doc: value,
+        extensions: [
+          highlightSpecialChars(),
+          history(),
+          drawSelection(),
+          dropCursor(),
+          EditorState.allowMultipleSelections.of(true),
+          indentOnInput(),
+          syntaxHighlighting(lightThemeHighlightingStyle, { fallback: true }),
+          bracketMatching(),
+          closeBrackets(),
+          autocompletion(),
+          // rectangularSelection(),
+          // crosshairCursor(),
+          highlightSelectionMatches({
+            wholeWords: true,
+            highlightWordAroundCursor: false, // Weird on numbers
+          }),
+          keymap.of([
+            ...closeBracketsKeymap,
+            ...defaultKeymap,
+            ...searchKeymap,
+            ...historyKeymap,
+            ...foldKeymap,
+            ...completionKeymap,
+            ...lintKeymap,
+            indentWithTab,
+          ]),
+          compGutter.of([]),
+          compUpdateListener.of([]),
+          compSubmitListener.of([]),
+          compTheme.of([]),
+          languageSupport,
+        ],
+      }),
+    []
   );
+
+  useEffect(() => {
+    if (editor.current != null && state != null) {
+      const view = new EditorView({ state, parent: editor.current });
+      editorView.current = view;
+
+      return () => {
+        view.destroy();
+      };
+    }
+  }, [state]);
+
+  useEffect(() => {
+    editorView.current?.dispatch({
+      effects: compGutter.reconfigure(
```

Comment:
`compGutter` is global, does this mean that there can't be two CodeEditors rendered together, one with gutter and other without it? If so, `compGutter` should be moved inside the component with `useMemo`.

Same concern for the following useEffect calls.

Reply:
comp**, and other compartments are just keys inside state. They won't interfere with other instances.